server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -605,9 +606,26 @@ private void doDeleteShardSnapshots(SnapshotId snapshotId, long repositoryStateI
             // Write the new repository data first (with the removed snapshot), using no shard generations
             final RepositoryData updatedRepoData = repositoryData.removeSnapshot(snapshotId, ShardGenerations.EMPTY);
             writeIndexGen(updatedRepoData, repositoryStateId, false, Function.identity(), ActionListener.wrap(v -> {
-                // Run unreferenced blobs cleanup in parallel to shard-level snapshot deletion
-                final ActionListener<Void> afterCleanupsListener =
-                    new GroupedActionListener<>(ActionListener.wrap(() -> listener.onResponse(null)), 2);
+                // Run unreferenced blobs cleanup in parallel to shard-level snapshot deletion and then apply the state transformer to the
+                // cluster state. When not using shard-generations, we must do this after doing all the shard folder updates because
+                // the repository is not safe for concurrent writes until all shard folders have been updated
+                final ActionListener<Void> afterCleanupsListener = new GroupedActionListener<>(ActionListener.wrap(() ->
```

Comment:
No great to have this here, but I still liked it better than handling this BwC upstream in the `SnapshotsService`